plat/common/include/x86/apic.h

```diff
@@ -44,30 +44,43 @@
 
 #include <errno.h>
 
-static inline int x2apic_enable(void)
+static inline int apic_enable(void)
 {
-	__u32 eax, ebx, ecx, edx;
+	__u32 eax, ebx __maybe_unused, ecx __maybe_unused, edx;
 
-	/* Check for x2APIC support */
-	cpuid(1, 0, &eax, &ebx, &ecx, &edx);
-	if (!(ecx & X86_CPUID1_ECX_x2APIC))
-		return -ENOTSUP;
-
-	/* Check if APIC is active */
+	/*
+	 * We expect to be started in xAPIC mode. This means firmware has not
+	 * irreversibly disabled the APIC and not yet enabled x2APIC mode.
+	 */
 	rdmsr(APIC_MSR_BASE, &eax, &edx);
-	if (!(eax & APIC_BASE_EN))
-		return -ENOTSUP;
+	UK_BUGON(!(eax & APIC_BASE_EN) || (eax & APIC_BASE_EXTD));
+
+	/* Set APIC software enable flag */
+	eax = ioreg_read32(APIC_MMIO_SVR);
+	eax |= APIC_SVR_EN;
+	ioreg_write32(APIC_MMIO_SVR, eax);
+
+	/* Enable "virtual wire mode" to receive PIC interrupts */
+	eax = APIC_LVT_TRIGGER_MODE_LEVEL | APIC_LVT_DELIVERY_MODE_EXTINT;
+	ioreg_write32(APIC_MMIO_LINT0, eax);
 
-	/* Switch to x2APIC mode */
-	eax |= APIC_BASE_EXTD;
-	wrmsr(APIC_MSR_BASE, eax, edx);
+	eax = APIC_LVT_TRIGGER_MODE_LEVEL | APIC_LVT_DELIVERY_MODE_NMI;
+	ioreg_write32(APIC_MMIO_LINT1, eax);
 
-	/* Set APIC software enable flag if necessary */
-	rdmsr(APIC_MSR_SVR, &eax, &edx);
-	if ((eax & APIC_SVR_EN) == 0) {
-		eax |= APIC_SVR_EN;
-		wrmsr(APIC_MSR_SVR, eax, edx);
+#ifdef CONFIG_HAVE_SMP
```

Comment:
this means a `!CONFIG_HAVE_SMP` build will succeed in `apic_enable` instead of failing with `-ENOTSUP` like before

Reply:
`!CONFIG_HAVE_SMP` did not call into `apic_enable` previously, or am I missing something here?

Reply:
you're right, I forgot about that